utils/driverless.c

```diff
@@ -59,6 +59,142 @@ convert_to_port(char *a)
   return (port);
 }
 
+/*
+ * 'copy_output()' - copy ippfind output data .
+ */
+
+static int				/* O - 0 on success, -1 on error */
+copy_output(int read_ipp_fd,		/* I - IPP file descriptor */
+             int read_ipps_fd,		/* I - IPPS file descriptor */
+             cups_array_t  *service_uri_list_ipps, /* I - IPPS CUPS Array */
+             cups_array_t  *service_uri_list_ipp) /* I - IPP CUPS Array */
+{
+  int		nfds;			/* Maximum file descriptor value + 1 */
+  fd_set	input;			/* Input set for reading */
+  struct timeval timeout;		/* Timeout for read... */
+  cups_file_t *fp;
+  int bytes;
+  char *ptr,
+        buffer[MAX_OUTPUT_LEN],	/* Copy buffer */
+        *ippfind_output;
+  
+ /*
+  * Figure out the maximum file descriptor value to use with select()...
+  */
+  nfds = (read_ipp_fd > read_ipps_fd ? read_ipp_fd : read_ipps_fd) + 1;
+
+ /*
+  * Now loop until we are out of data from read_ipp_fd and read_ipps_fd...
+  */
+
+  for (;;)
+  {
+   /*
+    * Use select() to determine whether we have data to copy around...
+    */
+
+    FD_ZERO(&input);
+    FD_SET(read_ipp_fd, &input);
+    FD_SET(read_ipps_fd, &input);
+    timeout.tv_sec  = 2;
+    timeout.tv_usec = 0;
+
+    if (select(nfds, &input, NULL, NULL, &timeout) < 0)
+      return (-1);
+    if (!FD_ISSET(read_ipp_fd, &input) && !FD_ISSET(read_ipps_fd, &input))
+    {
+      return (0);
+    } 
+
+    
+    if(FD_ISSET(read_ipps_fd, &input))
+    {
+      dup2(read_ipps_fd, 0);
+
+      fp = cupsFileStdin();
+
+     if ((bytes = cupsFileGetLine(fp, buffer, sizeof(buffer))) > 0) 
+      {
+        ippfind_output = (char *)malloc(MAX_OUTPUT_LEN*(sizeof(char)));
+        ptr = buffer;
+        while (ptr && !isalnum(*ptr & 255)) ptr ++;
+
+        if ((!strncasecmp(ptr, "ipps", 4) && ptr[4] == '\t'))
+        {
+          ptr += 4;
+          *ptr = '\0';
+          ptr ++;
+        } else
+          continue;
+        snprintf(ippfind_output, MAX_OUTPUT_LEN, "%s", ptr);
+        cupsArrayAdd(service_uri_list_ipps,ippfind_output);
+      }
+      else if (bytes < 0)
+      {
+        /*
+        * Read error - bail if we don't see EAGAIN or EINTR...
+        */
+        if (errno != EAGAIN && errno != EINTR)
+        {
+          perror("ERROR: Unable to read print data");
+          return (-1);
+        }
+      }
+      else
+      {
+        /*
+        * End of file, return...
+        */
+        return (0);
+      }
+      
+    }
+
+    if(FD_ISSET(read_ipp_fd, &input))
+    {
+      dup2(read_ipp_fd, 0);
+      fp = cupsFileStdin();
+      if ((bytes = cupsFileGetLine(fp, buffer, sizeof(buffer))) > 0) 
+      {
+        ippfind_output = (char *)malloc(MAX_OUTPUT_LEN*(sizeof(char)));
```

Comment:
@nidhijainpnp please free the memory you allocate.

Reply:
Hi,
I have to remove those two lines 
```
free(ippfind_output);
ippfind_output = NULL;
```
because it gives an error on testing:
```
nidhi@nidhi-X510UNR:~$ driverless 
Segmentation fault (core dumped)
```
Could you please confirm me if you are getting the same error on adding those two lines (freeing up the memory by `ippfind_output`)?

Reply:
Hi @nidhijainpnp ,

hmm, I don't get any segfault if I add the lines - with and without any driverless device on the local network. It would be good to see a segfault backtrace, then we can be able to figure out where the error is. 

The interesting is ```valgrind``` is not capable to find any leaks - my guess it is due multi-process - but if you create a minimal test case:

```
#include <stdlib.h>
#include <stdio.h>
#include <cups/array.h>
#include <string.h>


int main()
{

  char *i = (char*)malloc(10*sizeof(char));
  memset(i, 0, 10);

  cups_array_t *a = cupsArrayNew(NULL, NULL);

  cupsArrayAdd(a, i); 

  cupsArrayDelete(a);

  return 0;
}
```

, compile it with ```gcc -g -O0 -std=gnu11 -pedantic -Werror -Wall -o <output_binary> <source_file> -lcups``` and run ```valgrind --leak-check=full ./<output_binary>```, you will see a report for a leaking element.

If you look into ```cups/array.c``` of CUPS project, you see:
```
/*
 * 'cupsArrayDelete()' - Free all memory used by the array.
 *
 * The caller is responsible for freeing the memory used by the
 * elements themselves.
```

And ```service_uri``` is allocated on the heap too, so it should be freed.

Reply:
@nidhijainpnp you can pull the newest pull request which puts those lines back instead of manual copying.

Reply:
And I see you call:

```
nidhi@nidhi-X510UNR:~$ driverless 
```

Do you compile cups-filters master branch and install it via ```sudo make install``` or install the newest changes by other way?
Because IIUC this way you call the system ```driverless``` binary:

```
$ which driverless
/usr/bin/driverless
```

instead of ```driverless``` binary you built within project repository:

```
$ pwd
/home/zdohnal/repo_upstream/cups-filters
$ ls -lah driverless
-rwxrwxr-x. 1 zdohnal zdohnal 7.7K Sep  7 07:29 driverless
$ which ./driverless
./driverless
```

So the older ```driverless``` binary from system is called, unless you install the newest ```driverless``` from the project before.